### PR TITLE
fix: set access_token & refresh_token in cookies 

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
* to start the ball rolling for #279 
* sets access_token and refresh_token as individual cookies

## Steps to test
1. Add this to `.env` file
```
GOTRUE_COOKIE_KEY="sb"
GOTRUE_COOKIE_DURATION="100"
```
2. Update `GOTRUE_MAILER_AUTOCONFIRM=true` in the `.env` file
3. Signup a new user
```
curl -X POST "http://localhost:9999/signup" -H "Content-Type: application/json" -d '{"email": "user@example.com", "password": "secret"}'
```
4. Take note of the refresh token returned above.
5. Try to get a new access token & refresh token
```
curl -X POST 'http://localhost:9999/token?grant_type=refresh_token' \
-d '{"refresh_token": "refresh_token_in_4"}' -H "Content-type: application/json" -H "x-use-cookie: some-value" -v
```

## Expected Output
```
...
* Connected to localhost (127.0.0.1) port 9999 (#0)
> POST /token?grant_type=refresh_token HTTP/1.1
> Host: localhost:9999
> User-Agent: curl/7.77.0
> Accept: */*
> Content-type: application/json
> x-use-cookie: some-value
> Content-Length: 43
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json
< Set-Cookie: sb_accesstoken=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...; Path=/; Expires=Thu, 13 Jan 2022 02:51:16 GMT; Max-Age=10; HttpOnly; Secure
< Set-Cookie: sb_refreshtoken=R43XWbVOIvf3RpfK6A1qPg; Path=/; Expires=Thu, 13 Jan 2022 02:51:16 GMT; Max-Age=10; HttpOnly; Secure
< Vary: Origin
< Date: Thu, 13 Jan 2022 02:51:06 GMT
< Content-Length: 1353
< 
* Connection #0 to host localhost left intact
...
```

### CODE REVIEW

**Comment by kangmingtay** [comment]
Not too sure if we should remove [this check](https://github.com/supabase/gotrue/blob/master/api/token.go#L295) because of the following:
1. Seems like we should still check the cookie duration and set it if it's not expired
2. Should we also require the client to set the `x-use-cookie` as a header in the request instead of always setting the cookie?
3. [Netlify's use case for setting cookies](https://github.com/netlify/gotrue/pull/110#issue-261315439)

File: api/verify.go [hunk 1]
```diff
@@ -126,7 +126,7 @@ func (a *API) Verify(w http.ResponseWriter, r *http.Request) error {
 		}
 
 		if cookie != "" && config.Cookie.Duration > 0 {
-			if terr = a.setCookieToken(config, token.Token, cookie == useSessionCookie, w); terr != nil {
+			if terr = a.setCookieTokens(config, token, cookie == useSessionCookie, w); terr != nil {
```
Comment (dthyresson) [inline api/verify.go:129]: @kangmingtay based on the conversation last night with @thorwebdev, we always want to set the cookie and Thor suggested removing this check (and same in token, and signup).

Is this check still needed?

Comment (dthyresson) [reply api/verify.go:129]: Sorry, line is 128 -- the check on cookie and duration.

File: api/token.go [hunk 2]
```diff
@@ -200,7 +200,7 @@ func (a *API) ResourceOwnerPasswordGrant(ctx context.Context, w http.ResponseWri
 		}
 
 		if cookie != "" && config.Cookie.Duration > 0 {
```
Comment (dthyresson) [inline api/token.go:202]: Based on the conversation last night with @thorwebdev, we always want to set the cookie and Thor suggested removing this check. Is this check needed?

File: api/signup.go [hunk 3]
```diff
@@ -205,7 +205,7 @@ func (a *API) Signup(w http.ResponseWriter, r *http.Request) error {
 			}
 
 			if cookie != "" && config.Cookie.Duration > 0 {
```
Comment (dthyresson) [inline api/signup.go:207]: Based on the conversation last night with @thorwebdev, we always want to set the cookie and Thor suggested removing this check. Is this check needed?

File: api/token.go [hunk 4]
```diff
@@ -535,10 +548,18 @@ func (a *API) setCookieToken(config *conf.Configuration, tokenString string, ses
...
-	config := getConfig(ctx)
+func (a *API) clearCookieTokens(config *conf.Configuration, w http.ResponseWriter) {
+	a.clearCookieToken(config, "access_token", w)
+	a.clearCookieToken(config, "refresh_token", w)
+}
+
+func (a *API) clearCookieToken(config *conf.Configuration, name string, w http.ResponseWriter) {
+	cookieName := config.Cookie.Key
+	if name != "" {
+		cookieName += "_" + name
```
Comment (dthyresson) [inline api/token.go:559]: Using `_` as the delimiter might in the future make it hard to parse given `supabase_access_token` the suffix also has `_` so you have to split and pop off the first vs just split, but not a big deal. hold be fine unless one wants a more clear delimiter. 

> A <cookie-name> can contain any US-ASCII characters except for: the control character, space, or a tab. It also must not contain a separator characters like the following: ( ) < > @ , ; : \ " / [ ] ? = { }.

File: api/token.go [hunk 5]
```diff
@@ -517,10 +519,21 @@ func (a *API) issueRefreshToken(ctx context.Context, conn *storage.Connection, u
...
+	}
 	exp := time.Second * time.Duration(config.Cookie.Duration)
 	cookie := &http.Cookie{
-		Name:     config.Cookie.Key,
+		Name:     cookieName,
 		Value:    tokenString,
 		Secure:   true,
```
Comment (dthyresson) [inline api/token.go:538]: Something to note and test is that

> Insecure sites (http:) cannot set cookies with the Secure attribute (since Chrome 52 and Firefox 52). For Firefox, the https: requirements are ignored when the Secure attribute is set by localhost (since Firefox 75).

So, will want to test what the cookie is in local testing with just http.

File: api/token.go [hunk 6]
```diff
@@ -517,10 +519,21 @@ func (a *API) issueRefreshToken(ctx context.Context, conn *storage.Connection, u
...
+// setCookieTokens sets the access_token & refresh_token in the cookies
+func (a *API) setCookieTokens(config *conf.Configuration, token *AccessTokenResponse, session bool, w http.ResponseWriter) error {
+	_ = a.setCookieToken(config, "access_token", token.Token, session, w)
+	_ = a.setCookieToken(config, "refresh_token", token.RefreshToken, session, w)
+	return nil
+}
+
+func (a *API) setCookieToken(config *conf.Configuration, name string, tokenString string, session bool, w http.ResponseWriter) error {
+	cookieName := config.Cookie.Key
+	if name != "" {
```
Comment (dthyresson) [inline api/token.go:531]: Are we concerned that if `name` is "" then

>    If the user agent receives a new cookie with the same cookie-name,
   domain-value, and path-value as a cookie that it has already stored,
   the existing cookie is evicted and replaced with the new cookie.
   Notice that servers can delete cookies by sending the user agent a
   new cookie with an Expires attribute with a value in the past.

https://datatracker.ietf.org/doc/html/rfc6265#page-8

Which means a refresh token could be set in the access_token. Well, not in the code case, but if elsewhere this function is used.

Perhaps `name` **must** be a required parameter value?

**Review by dthyresson — COMMENTED** [review]
Left feedback, but looks like this is a great first step. 👍

File: api/token.go (hunk 4 above)
Comment (kangmingtay) [reply api/token.go:559]: hmm that's true but i can't think of a use case where it will be require to parse the cookie name? anyway, i can rename it to key_accesstoken & key_refreshtoken

File: api/token.go (hunk 6 above)
Comment (kangmingtay) [reply api/token.go:531]: good catch, ok will change it to if the name isn't specified, we don't set the cookie

File: api/verify.go (hunk 1 above)
Comment (thorwebdev) [reply api/verify.go:129]: @kangmingtay we'll need to remove it as it's impossible to provide the `x-use-cookie` header to the `/verify` route as it's called via the oauth redirect and not the gotrue-js client. I don't think there's any harm in always sending the cookie header, so I think we should remove.

File: api/token.go (hunk 5 above)
Comment (thorwebdev) [reply api/token.go:540]: It's possible on localhost: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies
image

File: api/token.go [hunk 7]
```diff
@@ -517,10 +520,22 @@ func (a *API) issueRefreshToken(ctx context.Context, conn *storage.Connection, u
 	}, nil
 }
 
-func (a *API) setCookieToken(config *conf.Configuration, tokenString string, session bool, w http.ResponseWriter) error {
+// setCookieTokens sets the access_token & refresh_token in the cookies
+func (a *API) setCookieTokens(config *conf.Configuration, token *AccessTokenResponse, session bool, w http.ResponseWriter) error {
+	// don't need to catch error here since we always set the cookie name
+	_ = a.setCookieToken(config, "accesstoken", token.Token, session, w)
```
Comment (thorwebdev) [inline api/token.go:526]:
```suggestion
	_ = a.setCookieToken(config, "access_token", token.Token, session, w)
```

File: api/token.go [hunk 8]
```diff
@@ -517,10 +520,22 @@ func (a *API) issueRefreshToken(ctx context.Context, conn *storage.Connection, u
...
+	_ = a.setCookieToken(config, "refreshtoken", token.RefreshToken, session, w)
```
Comment (thorwebdev) [inline api/token.go:527]:
```suggestion
	_ = a.setCookieToken(config, "refresh_token", token.RefreshToken, session, w)
```

File: api/token.go [hunk 9]
```diff
@@ -517,10 +520,22 @@ func (a *API) issueRefreshToken(ctx context.Context, conn *storage.Connection, u
...
+	return nil
+}
+
+func (a *API) setCookieToken(config *conf.Configuration, name string, tokenString string, session bool, w http.ResponseWriter) error {
+	if name == "" {
+		return errors.New("Failed to set cookie, invalid name")
+	}
+	cookieName := config.Cookie.Key + "_" + name
```
Comment (thorwebdev) [inline api/token.go:535]:
```suggestion
	cookieName := config.Cookie.Key + ":" + name
```

File: api/token.go [hunk 10]
```diff
@@ -535,10 +550,18 @@ func (a *API) setCookieToken(config *conf.Configuration, tokenString string, ses
 	return nil
 }
 
-func (a *API) clearCookieToken(ctx context.Context, w http.ResponseWriter) {
-	config := getConfig(ctx)
+func (a *API) clearCookieTokens(config *conf.Configuration, w http.ResponseWriter) {
+	a.clearCookieToken(config, "accesstoken", w)
```
Comment (thorwebdev) [inline api/token.go:554]:
```suggestion
	a.clearCookieToken(config, "access_token", w)
```

File: api/token.go [hunk 11]
```diff
@@ -535,10 +550,18 @@ func (a *API) setCookieToken(config *conf.Configuration, tokenString string, ses
...
+	a.clearCookieToken(config, "refreshtoken", w)
```
Comment (thorwebdev) [inline api/token.go:555]:
```suggestion
	a.clearCookieToken(config, "refresh_token", w)
```

File: api/token.go [hunk 12]
```diff
@@ -535,10 +550,18 @@ func (a *API) setCookieToken(config *conf.Configuration, tokenString string, ses
...
+}
+
+func (a *API) clearCookieToken(config *conf.Configuration, name string, w http.ResponseWriter) {
+	cookieName := config.Cookie.Key
+	if name != "" {
+		cookieName += "_" + name
```
Comment (thorwebdev) [inline api/token.go:561]:
```suggestion
		cookieName += ":" + name
```

File: api/token.go (hunk 4 above)
Comment (thorwebdev) [reply api/token.go:561]: I have a preference for `sb:access_token` and `sb:refresh_token`, wdyt?

**Review by thorwebdev — CHANGES_REQUESTED** [review]

File: api/token.go (hunk 4 above)
Comment (kangmingtay) [reply api/token.go:561]: i think the cookie name can't contain the ":" separator (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#attributes)

Comment (thorwebdev) [reply api/token.go:561]: Hm, strange, I'm using `sb:token` in gotrue-js and it works fine, and IIRC it was because auth0 was doing that. Then can de do `sb.access_token` instead?

File: api/token.go (hunk 2 above)
Comment (dthyresson) [reply api/token.go:203]: Sounds like from @thorwebdev in the above comment, the duration and cookie check can be removed here, too.

File: api/token.go (hunk 4 above)
Comment (dthyresson) [reply api/token.go:561]: I also see cookie name examples used `-`.

Perhaps:

`sb-access-token`

After some thought, the parsing the delimiter isn't all that critical.

And the `-` follows conventions like `x-use-cookie`

File: api/token.go (hunk 8 above)
Comment (dthyresson) [reply api/token.go:527]: Maybe `sb-refresh-token`?

File: api/token.go (hunk 10 above)
Comment (dthyresson) [reply api/token.go:554]: Maybe `sb-access-token`?

Or even

`Maybe `supabase-refresh-token`? For clarity?

**Comment by dthyresson** [comment]
Think it jut comes down to naming the cookies.

My votes are for:

* sb-access-token
* sb-refresh-token

or

* supabase-access-token
* supabase-refresh-token

But defer to @thorwebdev

File: api/token.go (hunk 2 above)
Comment (kangmingtay) [reply api/token.go:203]: I only removed it from the /verify endpoint because we can't pre-emptively set the header for GET requests. The /token endpoint is a POST request which means we can set the x-use-cookie header always.

Comment (thorwebdev) [reply api/token.go:203]: Yep, I think that's fine. I do think there isn't any harm in always setting the cookie across the board though if we wanted to align it across routes.

**Review by thorwebdev — APPROVED** [review]
Feel free to make a decision on the cookie naming 👍
